### ISSUE TITLE
CYCLE operation bug?

### ISSUE THREAD

**Issue by bmyersq** [martinisgroup/servers#151]
We ran into what seems to be a bug.

Our waveform is

waveform = np.zeros(16384)
waveform[0:2000] = 1

and is then converted to bytes for the sram.

For the jump table, we have

fpga.jump_table_add_entry('NOP', 16)
fpga.jump_table_add_entry('CYCLE', (5000, 0, 1, 0))
fpga.jump_table_add_entry('END', 6000)
fpga.jump_table_set_counter([5, 0, 0, 0])

the result is shown on first figure.

![figure1](https://cloud.githubusercontent.com/assets/12399363/7636839/90bac5c2-fa1e-11e4-8567-cc9cab106921.png)

If we set

fpga.jump_table_add_entry('NOP', 80)
fpga.jump_table_add_entry('CYCLE', (5000, 0, 1, 0))
fpga.jump_table_add_entry('END', 6000)
fpga.jump_table_set_counter([5, 0, 0, 0])

the result is shown on second figure.

![figure2](https://cloud.githubusercontent.com/assets/12399363/7636894/e2fa41b4-fa1e-11e4-9cbc-c049f0a2c71e.png)

Basically, it seems that if the fromAddress of the 'NOP' operation is too small, the 'CYCLE' operation doesn't work as expected.

**Comment by pomalley** [comment]
Yes--there is a hardware limitation that from addresses must be spaced at least 2 cells from each other. The from address of the NOP operation in the first example is 16 ns, which means cell 4. The convention is to subtract 2 from all from addresses, so the 4 becomes a 2 which is too close to the start address of 0. So we get undefined behavior, which you see here.

So this isn't a bug with cycle, but rather a hardware limitation.

How can we better document this?

**Comment by bmyersq** [comment]
We use a lot the file JumpTableIntro.md to understand how to run the board, it might be useful to document this in that file. Maybe also add a comment on the jump_table_add_entry documentation? Or raise an error when fromAddresses get too close?

Thanks for help.

**Comment by maffoo** [comment]
The server should raise an error if you try to run something like this, that is, it should know about and enforce these hardware constraints. I agree documentation is good, but we shouldn't rely on users reading the docs and only submitting valid commands. @pomalley, would it be difficult to make this happen?

**Comment by DanielSank** [comment]
+1 for code that helps you.

**Comment by pomalley** [comment]
OK, see the referenced commit and the previous one for code and documentation fixes.

**Comment by DanielSank** [comment]
@pomalley I'm pretty sure we can close this.

**Comment by pomalley** [comment]
Sure thing.
